torch/lib/TensorLib/templates/TensorDerived.cpp

```diff
@@ -0,0 +1,67 @@
+#include "TensorLib/${Tensor}.h"
+#include "TensorLib/HalfConvert.h"
+
+//sketch:
+
+/*
+
+class Type {
+  // ctor is configurable as well
+  Tensor new(...) {}
+  // all  methods and pure functions, virtually dispatched
+  virtual Tensor add(Tensor a, Tensor b);
+
+  // no Tensor arguments but virtually dispatched on type
+  virtual Tensor linspace(int a, int b);
+  //
+}
+
+// base class of derived tensors
+class TensorImpl {
+  Type * type_; //non-virtual so that we don't double indirect unnecessarily ...
+
+  // not virtual - Type * handles the virtual dispatch
+  Tensor add(Tensor b) {
+    type_->add(self,b);
+  }
+}
```

Comment:
The only thing I don't understand is why do we need three levels of indirection (`Tensor -> TensorImpl -> Type`). It seems that everything `Type` does could be handled by `TensorImpl`? Also, a very useful thing in context of e.g. CUDA tensor is that methods called on a tensor object automatically switch the device, but this would mean that you need a per-device type.